Add documentation about the different base and base_link frames (#286)
UniversalRobots/Universal_Robots_ROS2_Description: doc/robot_frames.rst

Robot Frames
============

The URDF model uses multiple frames in the robot's base. This page describes why that is and what

> the purpose of each of that frames is. In particular, there are three frames representing the

Reviewer: ```suggestion
the purpose of each of those frames is. In particular, there are three frames representing the
```

Reviewer: nit: I'm also not sure *representing the robot's base* is the best way to describe what's going on.

Author: You mean because of the term "base" being the same as the `base` frame which has a special meaning?

robot's base:

> - ``base_link`` - This serves as the root link if the kinematic chain. It follows `REP-103

Reviewer: ```suggestion
- ``base_link`` - This serves as the root link of the kinematic chain. It follows `REP-103
```

>   <https://ros.org/reps/rep-0103.html>`_ , where "forward" for the ``base_link`` is defined as the

Reviewer: ```suggestion
  <https://ros.org/reps/rep-0103.html>`_ , where *forward* for the ``base_link`` is defined as the
```

>   direction of the arm pointing to for an all-zero joint configuration.

Reviewer: point to where? :)

  .. figure:: frames/base_link.png
     :alt: The robot with an all-zeros joint configuration showing its base_link frame

>      The robot with an all-zeros joint configuration showing its base_link frame

Reviewer: ```suggestion
     The robot with an all-zeros joint configuration showing its ``base_link`` frame
```

?

Not sure figure captions support this

> - ``base`` - This is the frame that is used by the robot controller to represent the robot's base.

Reviewer: Does *base* refer to the physical base (ie: the link) or the `Base` frame?

The idea (of the REP in any case) would be the latter (ie: `Base` coordinate frame).

  It has the same position as ``base_link`` but is rotated by 180 degrees around the Z-axis.
  This follows the `REP proposal 199 <https://gavanderhoorn.github.io/rep/rep-0199.html>`_ with
  respect to the frame ``base`` and ``tool0``.

  Any lookup of ``tool0`` w.r.t. ``base`` will yield the same result as the pose shown on the teach

>   pendant when ``base`` is selected as a reference there and the tool is set to "all-zero" being in

Reviewer: Isn't it called `Base` (capital B) on the TP?

Author: Not on PolyScope X.

  the center of the flange.

  .. figure:: frames/base.png
     :alt: The robot with an all-zeros joint configuration showing its base frame

>      The robot with an all-zeros joint configuration showing its base frame

Reviewer: ```suggestion
     The robot with an all-zeros joint configuration showing its ``base`` frame
```

- ``base_link_inertia`` - Since some libraries such as KDL don't support inertia for the root link
  of a kinematic chain (see `ros/kdl_parser#27 <https://github.com/ros/kdl_parser/issues/27>`_), the ``base_link`` doesn't contain any meshes or inertia
  attached to it. Instead, those are attached to the ``base_link_inertia`` frame. This frame is in

>   fact rotated the same way as ``base``, which is for no particular reason.

Reviewer: > which is for no particular reason 

:confused:

This leads to the following kinematic chain:

.. code:: text

   base_link
   ├ base
   └ base_link_inertia
     └ shoulder_link
       └ upper_arm_link
         └ forearm_link
           └ wrist_1_link
             └ wrist_2_link
               └ wrist_3_link

>                  └ flange

Reviewer: might perhaps mention this frame as well and explain it's the intended `link` to attach end-effector (sub) `xacro:macro`s to?

                 └ tool0

The frame ``tool0`` is the tool frame as calculated using forward kinematics. If the robot has an

> all-zero tool configured, that is equivalent to the tool frame on the control box / teach pendant.

Reviewer: ```suggestion
all-zero tool configured, this should be equivalent to the tool frame on the control box / teach pendant.
```

.. note::

   When making TF lookups in the ROS system and comparing that to what the teach pendant shows,

>    please consider that the robot uses the ``base`` frame as reference, not ``base_link``. Also,

Reviewer: So the UR control box doesn't know anything about TF, or the structure of the URDF. The phrasing here sort-of implies it does, and explains the CB will use the frame called `base` instead of `base_link`.

If this is instead supposed to refer to concepts known/used by the UR CB, I'd perhaps use the name *Base*, as that's I believe the name of the *feature* the TCP pose will be relative to on the TP (if that feature is selected).

Author: Hm, that's a bit of a stretch. On PolyScope X the controller does indeed use TF and uses the term "frame" rather than "feature". 

![image](https://github.com/user-attachments/assets/9ef98f74-a8bc-4e31-a499-686fb4e5c6c2)

But we might have to make this clearer, yes. Thank you.

     The robot with an all-zeros joint configuration showing its ``base_link`` frame

- ``base`` - This is the frame that is used by the robot controller to represent the robot's *Base*
  feature (Polyscope 5) / *base* frame (PolyScope X).
  It has the same position as ``base_link`` but is rotated by 180 degrees around the Z-axis.
  This follows the `REP proposal 199 <https://gavanderhoorn.github.io/rep/rep-0199.html>`_ with
  respect to the frame ``base`` and ``tool0``.

  Any lookup of ``tool0`` w.r.t. ``base`` will yield the same result as the pose shown on the teach
  pendant when ``base`` is selected as a reference there and the tool is set to "all-zero" being in
  the center of the flange.

  .. figure:: frames/base.png
     :alt: The robot with an all-zeros joint configuration showing its base frame

     The robot with an all-zeros joint configuration showing its ``base`` frame

- ``base_link_inertia`` - Since some libraries such as KDL don't support inertia for the root link
  of a kinematic chain (see `ros/kdl_parser#27 <https://github.com/ros/kdl_parser/issues/27>`_),
  the ``base_link`` doesn't contain any meshes or inertia attached to it. Instead, those are

>   attached to the ``base_link_inertia`` frame. This frame is in rotated the same way as ``base``.

Reviewer 2: ```suggestion
  attached to the ``base_link_inertia`` frame. This frame is rotated in the same way as ``base``.
```
